:book: upgrade migration guide for v3 (#1916)
kubernetes-sigs/kubebuilder: docs/book/src/migration/migration_guide_v2tov3.md

this name is used in the scaffolds to create the name of your manager Pod and of the Namespace where the Manager is deployed by default.  

```bash
$ mkdir <my-project-name>
$ cd my-project-name
```

Now, we need to initialize a v3 project.  Before we do that, though, we'll need
to initialize a new go module if we're not on the `GOPATH`.

```bash
go mod init tutorial.kubebuilder.io/project
```

<aside class="note warning">
<h1>The module of your project can found in the `go.mod`:</h1>

```
module tutorial.kubebuilder.io/project
```

</aside>

Then, we can finish initializing the project with kubebuilder.

```bash
kubebuilder init --domain tutorial.kubebuilder.io
```

<aside class="note warning">
<h1>The domain of your project can be found in the PROJECT file:</h1>

```
...
domain: tutorial.kubebuilder.io
...
```
</aside>

## Migrate APIs and Controllers

Next, we'll re-scaffold out the API types and controllers. Since we want both,
we'll say yes to both the API and controller prompts when asked what parts we
want to scaffold:

```bash
kubebuilder create api --group batch --version v1 --kind CronJob
```

<aside class="note warning">

> <h1>How to still using `apiextensions.k8s.io/v1beta1` for CRDs?</h1>

Reviewer: ```suggestion
<h1>How to keep using `apiextensions.k8s.io/v1beta1` for CRDs?</h1>
```

From now on, the CRDs that will be created by the tool will be using the Kubenetes API version `apiextensions.k8s.io/v1`  by default, instead of `apiextensions.k8s.io/v1beta1` . So, if you would like to keep using the previous version, use the flag `--crd-version=v1beta` in the above command.  

</aside>

<aside class="note warning">
<h1>If you're using multiple groups</h1>

Please run `kubebuilder edit --multigroup=true` to enable multi-group support before migrating the APIs and controllers. Please see [this](/migration/multi-group.md) for more details.

</aside>

### Migrate the APIs

> Now, let's copy the API definition from `api/v1/cronjob_types.go` our old project to the new one. 

Reviewer: ```suggestion
Now, let's copy the API definition from `api/v1/cronjob_types.go` from our old project to the new one. 
```

### Migrate the Controllers

> Now, let's copy the controller code from `controllers/cronjob_controller.go` our old project to the new one and fix the following breaking change:  

Reviewer: ```suggestion
Now, let's copy the controller code from `controllers/cronjob_controller.go` from our old project to the new one and fix the following breaking change:  
```

 
Replace:

```go 
func (r *CronJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background() 
 	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

With:

```go 
func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

<aside class="note warning">
<h1>Controller-runtime version updated has breaking changes</h1>

Check [sigs.k8s.io/controller-runtime release docs from 0.7.0+ version]( https://github.com/kubernetes-sigs/controller-runtime/releases ) for breaking changes.

</aside>

## Migrate the Webhooks

<aside class="note warning">
<h1>Note</h1>

> If you don't have a webhook, you can skip this section.

Reviewer: ```suggestion
If you don't have any webhooks, you can skip this section.
```

</aside>

Now let's scaffold the webhooks for our CRD (CronJob). We'll need to run the
following command with the `--defaulting` and `--programmatic-validation` flags
(since our test project uses defaulting and validating webhooks):

```bash
kubebuilder create webhook --group batch --version v1 --kind CronJob --defaulting --programmatic-validation
```

<aside class="note warning">

> <h1>How to still using `apiextensions.k8s.io/v1beta1` for Webhooks?</h1>

Reviewer: ```suggestion
<h1>How to keep using `apiextensions.k8s.io/v1beta1` for Webhooks?</h1>
```

From now on, the Webhooks that will be created by the tool will be using by default the Kubernetes API version `apiextensions.k8s.io/v1` instead of `apiextensions.k8s.io/v1beta1`,  `admissionregistration.k8s.io/v1` instead of `admissionregistration.k8s.io/v1beta1` and the `cert-manager.io/v1` to replace `cert-manager.io/v1alpha2`. So, if you would like to keep using the previous version use the flag `--webhook-version=v1beta` in the above command.  
</aside>

> Now, let's copy the API definition from `api/v1/cronjob_webhook.go` our old project to the new one. 

Reviewer: ```suggestion
Now, let's copy the API definition from `api/v1/cronjob_webhook.go` from our old project to the new one. 
```

to install the required components.

The recommended way to migrate a v2 project is to create a new v3 project and
copy over the API and the reconciliation code. The conversion will end up with a
project that looks like a native v3 project. However, in some cases, it's
possible to do an in-place upgrade (i.e. reuse the v2 project layout, upgrading
controller-runtime and controller-tools.  

## Initialize a v3 Project

Create a new directory with the name of your project. Note that
this name is used in the scaffolds to create the name of your manager Pod and of the Namespace where the Manager is deployed by default.  

```bash
$ mkdir <my-project-name>
$ cd my-project-name
```

Now, we need to initialize a v3 project.  Before we do that, though, we'll need
to initialize a new go module if we're not on the `gopath`.

```bash
go mod init tutorial.kubebuilder.io/project`
```

<aside class="note warning">
<h1>Note</h1>
The module of your project can found in the `go.mod`:

```
module tutorial.kubebuilder.io/project
```

</aside>

Then, we can finish initializing the project with kubebuilder.

```bash
kubebuilder init --domain tutorial.kubebuilder.io
```

<aside class="note warning">

> <h1>Note</h1>

Author: I think I could solve that.

The domain of your project can be found in the PROJECT file:
```
...
domain: tutorial.kubebuilder.io
...
```
</aside>

## Migrate APIs and Controllers

Next, we'll re-scaffold out the API types and controllers. Since we want both,
we'll say yes to both the API and controller prompts when asked what parts we
want to scaffold:

```bash
kubebuilder create api --group batch --version v1 --kind CronJob
```

<aside class="note warning">
<h1>Note</h1>

> From now on, the CRDs that will be created by the tool will be using by default the Kubenetes API version `apiextensions.k8s.io/v1` instead of `apiextensions.k8s.io/v1beta1` . So, if you would like to keep using the previous version use the flag `--crd-version=v1beta` in the above command.  

Author: done

before continuing.

Please ensure you have followed the [installation guide](/quick-start.md#installation)
to install the required components.

The recommended way to migrate a v2 project is to create a new v3 project and
copy over the API and the reconciliation code. The conversion will end up with a
project that looks like a native v3 project. However, in some cases, it's
possible to do an in-place upgrade (i.e. reuse the v2 project layout, upgrading
controller-runtime and controller-tools.  

## Initialize a v3 Project

Create a new directory with the name of your project. Note that
this name is used in the scaffolds to create the name of your manager Pod and of the Namespace where the Manager is deployed by default.  

```bash
$ mkdir <my-project-name>
$ cd my-project-name
```

Now, we need to initialize a v3 project.  Before we do that, though, we'll need
to initialize a new go module if we're not on the `GOPATH`.

```bash
go mod init tutorial.kubebuilder.io/project
```

<aside class="note warning">
<h1>The module of your project can found in the `go.mod`:</h1>

```
module tutorial.kubebuilder.io/project
```

> </aside>

Author: Users sometimes does not know where found this info. the idea here is only made clear that.

Then, we can finish initializing the project with kubebuilder.

```bash
kubebuilder init --domain tutorial.kubebuilder.io
```

<aside class="note warning">
<h1>The domain of your project can be found in the PROJECT file:</h1>

```
...
domain: tutorial.kubebuilder.io
...
```
</aside>

## Migrate APIs and Controllers

Next, we'll re-scaffold out the API types and controllers. Since we want both,
we'll say yes to both the API and controller prompts when asked what parts we
want to scaffold:

```bash
kubebuilder create api --group batch --version v1 --kind CronJob
```

<aside class="note warning">
<h1>How to still keep `apiextensions.k8s.io/v1beta1` for CRDs?</h1>

> From now on, the CRDs that will be created by the tool will be using the Kubenetes API version `apiextensions.k8s.io/v1`  by default, instead of `apiextensions.k8s.io/v1beta1` . So, if you would like to keep using the previous version, use the flag `--crd-version=v1beta` in the above command.  

Reviewer 2: say which tool instead of just the tool:

```suggestion
From now on, the CRDs that will be created by controller-gen will be using the Kubenetes API version `apiextensions.k8s.io/v1`  by default, instead of `apiextensions.k8s.io/v1beta1` . So, if you would like to keep using the previous version, use the flag `--crd-version=v1beta` in the above command.  
```

Also, should that be `v1beta1`, not `v1beta`?

func (r *CronJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background() 
 	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

With:

```go 
func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

<aside class="note warning">
<h1>Controller-runtime version updated has breaking changes</h1>

Check [sigs.k8s.io/controller-runtime release docs from 0.7.0+ version]( https://github.com/kubernetes-sigs/controller-runtime/releases ) for breaking changes.

</aside>

## Migrate the Webhooks

<aside class="note warning">
<h1>Note</h1>

If you don't have any webhooks, you can skip this section.

</aside>

Now let's scaffold the webhooks for our CRD (CronJob). We'll need to run the
following command with the `--defaulting` and `--programmatic-validation` flags
(since our test project uses defaulting and validating webhooks):

```bash
kubebuilder create webhook --group batch --version v1 --kind CronJob --defaulting --programmatic-validation
```

<aside class="note warning">
<h1>How to keep using `apiextensions.k8s.io/v1beta1` for Webhooks?</h1>

> From now on, the Webhooks that will be created by Kubebuilder using by default the Kubernetes API version `apiextensions.k8s.io/v1` instead of `apiextensions.k8s.io/v1beta1`,  `admissionregistration.k8s.io/v1` instead of `admissionregistration.k8s.io/v1beta1` and the `cert-manager.io/v1` to replace `cert-manager.io/v1alpha2`. So, if you would like to keep using the previous version use the flag `--webhook-version=v1beta` in the above command.  

Author: if you use the latest version of webhook than the new cert-manager will be used otherwise it should scaffold the legacy one. If it is not working in this way is a bug.

<aside class="note">
<h1>Scaffolding both the API types and controllers</h1>

For this example, we are going to consider that we need to scaffold both the API types and the controllers, but remember that this depends on how you scaffolded them in your original project.

</aside>

```bash
kubebuilder create api --group batch --version v1 --kind CronJob
```

<aside class="note">
<h1>How to still keep `apiextensions.k8s.io/v1beta1` for CRDs?</h1>

From now on, the CRDs that will be created by controller-gen will be using the Kubenetes API version `apiextensions.k8s.io/v1`  by default, instead of `apiextensions.k8s.io/v1beta1` . So, if you would like to keep using the previous version, use the flag `--crd-version=v1beta1` in the above command.  

</aside>

### Migrate the APIs

<aside class="note">
<h1>If you're using multiple groups</h1>

Please run `kubebuilder edit --multigroup=true` to enable multi-group support before migrating the APIs and controllers. Please see [this](/migration/multi-group.md) for more details.

</aside>

Now, let's copy the API definition from `api/v1/<kind>_types.go` from our old project to the new one.

> This files have not been modified by the new plugin, so you should be able to replace your freshly scaffolded.

Reviewer 3: ```suggestion
This files have not been modified by the new plugin, so you should be able to replace your freshly scaffolded
```

```go 
func (r *CronJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background() 
 	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

With:

```go 
func (r *CronJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

<aside class="note warning">
<h1>Controller-runtime version updated has breaking changes</h1>

Check [sigs.k8s.io/controller-runtime release docs from 0.7.0+ version]( https://github.com/kubernetes-sigs/controller-runtime/releases ) for breaking changes.

</aside>

## Migrate the Webhooks

<aside class="note">
<h1>Skip</h1>

If you don't have any webhooks, you can skip this section.

</aside>

Now let's scaffold the webhooks for our CRD (CronJob). We'll need to run the
following command with the `--defaulting` and `--programmatic-validation` flags
(since our test project uses defaulting and validating webhooks):

```bash
kubebuilder create webhook --group batch --version v1 --kind CronJob --defaulting --programmatic-validation
```

<aside class="note">
<h1>How to keep using `apiextensions.k8s.io/v1beta1` for Webhooks?</h1>

> From now on, the Webhooks that will be created by Kubebuilder using by default the Kubernetes API version `apiextensions.k8s.io/v1` instead of `apiextensions.k8s.io/v1beta1`,  `admissionregistration.k8s.io/v1` instead of `admissionregistration.k8s.io/v1beta1` and the `cert-manager.io/v1` to replace `cert-manager.io/v1alpha2`. So, if you would like to keep using the previous version use the flag `--webhook-version=v1beta1` in the above command.  

Reviewer 3: Check which versions of K8s require this and add like the above.

project that looks like a native v3 project. However, in some cases, it's
possible to do an in-place upgrade (i.e. reuse the v2 project layout, upgrading
[controller-runtime][controller-runtime] and [controller-tools][controller-tools]).  

## Initialize a v3 Project

<aside class="note">
<h1>Project name</h1>

For the rest of this document, we are going to use `migration-project` as the project name and `tutorial.kubebuilder.io` as the domain. Please, select and use appropriate values for your case.

</aside>

Create a new directory with the name of your project. Note that
this name is used in the scaffolds to create the name of your manager Pod and of the Namespace where the Manager is deployed by default.  

```bash
$ mkdir migration-project-name
$ cd migration-project-name
```

Now, we need to initialize a v3 project.  Before we do that, though, we'll need
to initialize a new go module if we're not on the `GOPATH`. While technically this is
not needed inside `GOPATH`, it is still recommended.

```bash
go mod init tutorial.kubebuilder.io/migration-project
```
<aside class="note warning">
<h1> Migrating to Kubebuilder v3 while staying on the go/v2 plugin </h1>

> You can use `--plugins=go/v2` if you wish continue using "`Kubebuilder 2.x`" layout and avoid dealing with the breaking changes that will be faced because of the default upper versions which will be used now. See that the [controller-tools][controller-tools] v0.4.1 & [controller-runtime][controller-runtime] v0.8.0 are just used by default with the `go/v3` plugin layout. 

Reviewer 4: I'm in need of some clairvoyant clarification since I have been dealing with A LOT LATELY.  Does this mean that my code is no longer going to be stolen by hashicorp employees and passed off as their own work? ![image](https://user-images.githubusercontent.com/89703573/148107709-c654e8f1-51ae-4c4d-8b35-38310a2db662.png)

Or they were cleverly programmed bots impersonating people either way im exhausted of the misdirection, deliberately confusing naming schema designed to obtain a nefarious outcome, blatant disregard for decency and respect which was harassment to put it lightly as they repeatedly blocked me from resources I was seeking for help.   Code hidden in white space and redirects like the one that sent me to their page where a code immediately began executing a build of mine.  Ever since I told them I didn't want to sign their cla and had questions.  

I still have a question? What about all the work my Ai built for them? I made the Ai and they built out an astronomical amount of code.  Now what?  The why was I considered to be the hashicorp packer or was that a mistake from naming crossovers? 

I need clarity and thank everyone who is listening and not dismissing this like before the hundred times i requested assistance... please forgive my candor as I am in need of rest from having to constantly fix loopholes being abused...![image](https://user-images.githubusercontent.com/89703573/148109648-6fa21544-60fc-4f8a-ad8a-45310e97bae1.png)

</aside>

<aside class="note">

> <h1>The module of your project can found in the `go.mod`:</h1>

Reviewer 5: maybe say where this is? "in the 'go.mod' file at the root of your project"

```
module tutorial.kubebuilder.io/migration-project
```

</aside>

Then, we can finish initializing the project with kubebuilder.

```bash
kubebuilder init --domain tutorial.kubebuilder.io
```

<aside class="note">
<h1>The domain of your project can be found in the PROJECT file:</h1>

```yaml
...
domain: tutorial.kubebuilder.io
...
```
</aside>

## Migrate APIs and Controllers

Next, we'll re-scaffold out the API types and controllers.

<aside class="note">
<h1>Scaffolding both the API types and controllers</h1>

For this example, we are going to consider that we need to scaffold both the API types and the controllers, but remember that this depends on how you scaffolded them in your original project.

</aside>

```bash
kubebuilder create api --group batch --version v1 --kind CronJob
```

<aside class="note">
<h1>How to still keep `apiextensions.k8s.io/v1beta1` for CRDs?</h1>

From now on, the CRDs that will be created by controller-gen will be using the Kubernetes API version `apiextensions.k8s.io/v1`  by default, instead of `apiextensions.k8s.io/v1beta1`. 

> The `apiextensions.k8s.io/v1beta1` was deprecated in Kubernetes `1.16` and will be removed  in Kubernetes `1.22`.

Reviewer 6: ```suggestion
The `apiextensions.k8s.io/v1beta1` was deprecated in Kubernetes `1.16` and will be removed in Kubernetes `1.22`.
```

So, if you would like to keep using the previous version use the flag `--crd-version=v1beta1` in the above command which is only needed if you want your operator to support Kubernetes `1.15` and earlier.

</aside>

### Migrate the APIs

<aside class="note">
<h1>If you're using multiple groups</h1>

Please run `kubebuilder edit --multigroup=true` to enable multi-group support before migrating the APIs and controllers. Please see [this][multi-group] for more details.

</aside>

> Now, let's copy the API definition from `api/v1/<kind>_types.go` from our old project to the new one.

Reviewer 5: maybe "from `api/v1/<kind>_types.go` in our old project to the new one."?

These files have not been modified by the new plugin, so you should be able to replace your freshly scaffolded files by your old one. There may be some cosmetic changes. So you can choose to only copy the types themselves.

### Migrate the Controllers

> Now, let's migrate the controller code from `controllers/cronjob_controller.go` from our old project to the new one. There is a breaking change and there may be some cosmetic changes.

Reviewer 5: same as above

	log := r.Log.WithValues("cronjob", req.NamespacedName)
```

<aside class="note warning">
<h1>Controller-runtime version updated has breaking changes</h1>

Check [sigs.k8s.io/controller-runtime release docs from 0.7.0+ version][controller-runtime] for breaking changes.

</aside>

## Migrate the Webhooks

<aside class="note">
<h1>Skip</h1>

If you don't have any webhooks, you can skip this section.

</aside>

Now let's scaffold the webhooks for our CRD (CronJob). We'll need to run the
following command with the `--defaulting` and `--programmatic-validation` flags
(since our test project uses defaulting and validating webhooks):

```bash
kubebuilder create webhook --group batch --version v1 --kind CronJob --defaulting --programmatic-validation
```

<aside class="note">
<h1>How to keep using `apiextensions.k8s.io/v1beta1` for Webhooks?</h1>

From now on, the Webhooks that will be created by Kubebuilder using by default the Kubernetes API version `admissionregistration.k8s.io/v1` instead of `admissionregistration.k8s.io/v1beta1` and the `cert-manager.io/v1` to replace `cert-manager.io/v1alpha2`. 

> The `apiextensions/v1beta1` and `admissionregistration.k8s.io/v1beta1` was deprecated in Kubernetes `1.16` and will be removed  in Kubernetes `1.22`. If you use `apiextensions/v1` and `admissionregistration.k8s.io/v1` then you need to use `cert-manager.io/v1` which will be the API adopted per Kubebuilder CLI by default in this case.  

Reviewer 5: maybe "`apiextensions/v1beta1` and `admissionregistration.k8s.io/v1beta1` were"

> Note that, the API `cert-manager.io/v1alpha2` is not compatible with the latest Kubernetes APIs. 

Reviewer 5: don't need a comma here
